stream/src/main/java/org/apache/kafka/streaming/processor/internals/PartitionGroup.java

```diff
@@ -50,31 +53,45 @@ public int compare(RecordQueue queue1, RecordQueue queue2) {
             }
         });
 
-        totalBuffered = 0;
+        this.partitionQueues = partitionQueues;
+
+        this.totalBuffered = 0;
     }
 
     /**
-     * Get the next record from the partition with the lowest timestamp to be processed
+     * Get one record from the specified partition queue
      */
-    public StampedRecord nextRecord() {
-
-        // Get the partition with the lowest timestamp.
-        RecordQueue recordQueue = queuesByTime.poll();
+    public StampedRecord getRecord(RecordQueue queue) {
+        // get the first record from this queue.
+        StampedRecord record = queue.poll();
 
-        // Get the first record from this partition's queue.
-        StampedRecord record = recordQueue.next();
+        // update the partition's timestamp and re-order it against other partitions.
 
-        totalBuffered--;
+        queuesByTime.remove(queue);
```

Comment:
I have been pondering on this: the reason I added two functions nextQueue() and getRecord instead of just one function nextRecord is that the caller StreamTask needs to know both the queue and the record, and I do not want to have nextRecord returning a tuple. But the disadvantage of having two functions is that theoretically we cannot guarantee that caller will trigger getRecord right after nextQueue.